### ISSUE TITLE
gRPC September 2023 Issue

### ISSUE THREAD

**Issue by ggilmore** [sourcegraph/sourcegraph#55629]
# gRPC September 2023 Issue

Overall, the primary goal is to support the managed instance gRPC rollout and respond to any issues that arise from it. 

After this iteration, gRPC will be on by default in the 5.2.X release. In the 5.3.X release, gRPC will be the only option - the legacy REST codepaths will be removed.

**Note**: Even after the 5.2.X release, there will be a large number of follow up tasks:

- any follow up configuration changes that arise from the launch will self-hosted customers
- any tasks required to remove the legacy HTTP REST methods
- the syntax highlighter service still needs to be converted to use gRPC

## about this tracking issue 

**Since there aren't really fixed-length tasks for gRPC this iteration -  and more on going support, the tracking issue is represented as a list of priorities instead of a strict task lists.**

# priority 1: support gRPC managed instance rollout

gRPC is rolling out on al of our managed instances up through the end of August. See **[Cloud feature launch: gRPC#2228](https://www.notion.so/gRPC-September-2023-Issue-0e53a429fa8140289f0b8c6f87b9c40e?pvs=21)** for the full schedule. 

The cloud team has set up monitoring in such a way that we’ll be notified if any errors from the “internal error” logic that we added pop up on any of the instances. 

Our top priority is to respond to and resolve to any alerts that popup from these instances, or any other issues that the cloud team brings to our attention (or our customers let us know about). 

# priority 2: ongoing issue rundowns

There are two known classes of issues that we need to fix individual instances of. In general, there are no one sized fits all solutions to theses issues, we fix invidiual occurrences of the problem as they arise. 

In order to find new occurrences of these problems, we look at reports from:

- [sourcegraph.com](http://sourcegraph.com)’s gRPC error logs: [https://app.golinks.io/prod-logs-grpc-errors](https://app.golinks.io/prod-logs-grpc-errors)
- alerts-managed-instances (gRPC specific errors): [https://sourcegraph.slack.com/archives/C017SLJGA2Z](https://sourcegraph.slack.com/archives/C017SLJGA2Z)

## non-utf8 string issue

**Issue**: [gRPC error reporting indicates non-UTF-8 string fields are slipping through somehow](https://github.com/sourcegraph/sourcegraph/issues/52181)

**TL;DR:** Our conversion to gRPC has revealed errors due to non-UTF-8 strings in certain gRPC endpoints (`gitserver.v1.GitserverService` and `symbols.v1.SymbolsService`). This issue arises from inconsistencies in how the system handles strings and the assumption that strings are always UTF-8 encoded. The protobuf's `string` type only allows UTF-8 encoded strings, however, Go's string type does not enforce this, and some functions might not work correctly with non-UTF-8 strings. This is likely to be a broader application issue that gRPC's stricter requirements have revealed.

There isn’t a one-sized-fits-all type of solution to this issue. We take each individual occurrence of the issue (from [sourcegraph.com](http://sourcegraph.com)’s [error logs](https://console.cloud.google.com/logs/query;query=resource.type%3D%22k8s_container%22%0Aresource.labels.project_id%3D%22sourcegraph-dev%22%0Aresource.labels.location%3D%22us-central1-f%22%0Aresource.labels.cluster_name%3D%22cloud%22%0Aresource.labels.namespace_name%3D%22prod%22%0AjsonPayload.InstrumentationScope%3D~%22gRPC.internal.error.reporter%22%0A;cursorTimestamp=2023-08-06T02:26:37.786373604Z;duration=P2D?project=sourcegraph-dev) or from managed instance alerts) and come up with a fix for that unique occurrence. 

Fixes:

- [x]  [https://github.com/sourcegraph/sourcegraph/pull/55269](https://github.com/sourcegraph/sourcegraph/pull/55269)
- [x]  [https://github.com/sourcegraph/sourcegraph/pull/55275](https://github.com/sourcegraph/sourcegraph/pull/55275)
- [x]  [https://github.com/sourcegraph/sourcegraph/pull/55327](https://github.com/sourcegraph/sourcegraph/pull/55327)
- [x]  [https://github.com/sourcegraph/sourcegraph/pull/55369](https://github.com/sourcegraph/sourcegraph/pull/55369)
- [x]  [https://github.com/sourcegraph/sourcegraph/pull/55449](https://github.com/sourcegraph/sourcegraph/pull/55449)

## message size too large

gRPC wasn’t optimized for sending large messages. Quoting from [Microsoft’s “Performance Best Practices with gRPC” page](https://learn.microsoft.com/en-us/aspnet/core/grpc/performance)

> gRPC is a message-based RPC framework, which means:
> 
> - The entire message is loaded into memory before gRPC can send it.
> - When the message is received, the entire message is deserialized into memory.

As such, large messages can cause a lot of memory allocations for 
both the client and server. For this reason, the Go GRPC implementation 
has the following default limits

- [sending](https://pkg.go.dev/google.golang.org/grpc#MaxSendMsgSize): maximum ~2.1 GB (2^32-1 bytes) message
- [receiving](https://pkg.go.dev/google.golang.org/grpc#MaxRecvMsgSize): maximum 4 MB message

In the 5.1.X release, we’ve already implemented a workaround to be able to [tune the maximum size via environment variables](https://github.com/sourcegraph/sourcegraph/pull/52036/files). However, this is just a workaround while we work to resolve the problem (and can get rid of these environment variables). 

Fixes: 

- [x]  [https://github.com/sourcegraph/sourcegraph/pull/55242](https://github.com/sourcegraph/sourcegraph/pull/55242)
- [x]  [https://github.com/sourcegraph/sourcegraph/pull/55209](https://github.com/sourcegraph/sourcegraph/pull/55209)
- [x]  [https://github.com/sourcegraph/sourcegraph/pull/52208](https://github.com/sourcegraph/sourcegraph/pull/52208)

# priority 3: convert configuration and zoekt-webserver services to gRPC

- [x]  Switching over the configuration package to speak gRPC.

- [x]  [fix zoekt-webserver gRPC support](https://github.com/sourcegraph/sourcegraph/issues/53356): We tried converting zoekt-webserver to gRPC for the 5.1.X release, but we ran into last-minute issues that caused us to have to put it behind an additional feature flag (turning it off).

# priority 4: get test self-hosted customers to enable gRPC on their instances

Our next priority will be to chat with our delivery teams and ask them to connect us with some test customers to enable gRPC on their self-hosted instances. We will target friendly and responsive admins and perhaps ones that have odd networking setups (firewalls, service meshes).

We'll use these experiences to build out enablement materials ([like this](https://handbook.sourcegraph.com/departments/engineering/dev/tools/grpc/)) for the delivery / CS teams.

**Note**: **this is probably the task that will get cut if necessary**.

- If we decide to cut this, the trade-off here is between having those pre-emptive configuration conversations to head off those issues versus being very responsive to helping admins out after gRPC is on by default in the 5.2.x release. I am comfortable making this trade-off since I'm confident that the kinds of issues we'll see would be related to the customer's restrictive firewall configuration (and settings that they’ll need to change)

/cc @sourcegraph/source

**Comment by ggilmore** [comment]
This was delivered during the last release.
